File: problems/average-of-levels-bst.cc
#include <bits/stdc++.h>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  vector<double> averageOfLevels(TreeNode *root) {
    queue<TreeNode *> q;
    vector<double> result;

    q.push(root);

    while (!q.empty()) {
      long count = q.size();
      double sum = 0;

      for (long i = 0; i < count; i++) {
        TreeNode *curr = q.front();
        q.pop();
        sum += curr->val;

        if (curr->left != nullptr) {
          q.push(curr->left);
        }
        if (curr->right != nullptr) {
          q.push(curr->right);
        }
      }
      result.push_back(sum / count);
    }
  }
};

File: problems/average-of-levels-bst.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var averageOfLevels = function (root) {
  var result = [];
  var q = [];
  q.push(root);

  while (q.length) {
    var count = q.length;
    var sum = 0;

    for (let i = 0; i < count; i++) {
      var curr = q.shift();
      sum += curr.val;

      if (curr.left) q.push(curr.left);

      if (curr.right) q.push(curr.right);
    }
    result.push(sum / count);
  }
  return result;
};

const root = new TreeNode(
  3,
  new TreeNode(9, null, null),
  new TreeNode(20, new TreeNode(15, null, null), new TreeNode(7, null, null))
);

console.log(averageOfLevels(root));
